### PR TITLE
Finish support for hook priorities

### PR DESCRIPTION
**Pull request by rsmmr**

### CODE REVIEW

File: tests/hilti/types/function/hook-priorities.hlt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+module Bar {
+
+import hilti;
+import Foo;
+
+public function hook void Foo::f() &priority=-5 { hilti::print("f with priority -5"); }
+public function hook void Foo::f() &priority=15 { hilti::print("f with priority 15"); }
+
+
```
Comment (bbannier) [inline tests/hilti/types/function/hook-priorities.hlt:28]: nit: dead line

File: tests/hilti/types/function/hook-priorities.hlt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+@TEST-START-FILE foo.hlt
+
+module Foo {
+
+import hilti;
+
+declare public hook void f();
+
+public function hook void f() { hilti::print("f with priority 0"); }
```
Comment (bbannier) [inline tests/hilti/types/function/hook-priorities.hlt:12]: Could make this `f with default priority` to decouple from whatever we use there.

File: tests/spicy/types/unit/hooks-priorities.spicy [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# @TEST-EXEC:  echo 12345| spicy-driver -p A::tA %INPUT b.spicy >output
+# @TEST-EXEC:  btest-diff output
+
+module A;
+
+public type tA = unit {
+       a: bytes &size=5 {
+            print "A: prio 0", self.a;
```
Comment (bbannier) [inline tests/spicy/types/unit/hooks-priorities.spicy:8]: similarly: `default prio`

File: doc/programming/parsing.rst [hunk 4]
```diff
@@ -215,7 +215,7 @@ position. Example:
     :show-with: foo.spicy
 
 As you can see, ``x`` receives 6 bytes of input, which it then turns
-into three 16-bit integers.
+into three a-bit integers.
```
Comment (bbannier) [inline doc/programming/parsing.rst:218]: Why this change?

File: doc/programming/parsing.rst [hunk 5]
```diff
@@ -143,7 +143,7 @@ chapter.
 Unit fields always have a type. However, in some cases a field's type
 is not explicitly declared, but derived from what's being parsed. The
 main example of this is parsing a constant value: Instead of a type, a
-field can specify a constant of a parseable type. The field's type
+field can specify a constant of a parse able type. The field's type
```
Comment (bbannier) [inline doc/programming/parsing.rst:146]: I feel the old form was already correct.

File: doc/programming/parsing.rst [hunk 6]
```diff
@@ -314,7 +314,7 @@ constraints on their values. ``EXPR`` must a be boolean expression
 that will be evaluated after the parsing for the field has finished,
 but before any hooks execute. If ``EXPR`` returns ``False``, the
 parsing process will abort with an error, just as if the field had
-been unparseable in the first place (incl. executing any :ref:`%error
+been untraceable in the first place (incl. executing any :ref:`%error
```
Comment (bbannier) [inline doc/programming/parsing.rst:317]: Was this change intentional? I think I understood _unparseable, but _untraceable_ would appear here the only time in the user-searchable docs, and at least for me means something different.

File: doc/programming/parsing.rst [hunk 7]
```diff
@@ -469,6 +469,18 @@ There are three locations where hooks can be implemented:
...
+If multiple implementations are provided for the same hook, by default
+it remains undefined in which order they will execute. If a particular
+order is desired, you can specify priorities for your hook
+implementations::
+
+      on Foo::v priority=-5 { ... }
+      on Foo::v priority=5 { ... }
+
+Implementations then execute in order of their priorities: The lower a
+priority value, the earlier it will execute. If not specified, a
```
Comment (bbannier) [inline doc/programming/parsing.rst:481]: Ah I see, maybe disregard my comment from above on possibly leaving them unset by default is not the behavior we want. Do you have a specific use case for hooks with negative priority in mind? -- maybe something worth adding an example for.

File: hilti/toolchain/include/ast/attribute.h [hunk 8]
```diff
@@ -72,19 +73,40 @@ class Attribute : public NodeBase {
...
+            if ( ! hasValue() )
+                return result::Error(hilti::util::fmt("attribute '%s' requires a string", _tag));
+
             if ( auto e = value().tryAs<expression::Ctor>() )
                 if ( auto s = e->ctor().tryAs<ctor::String>() )
                     return s->value();
             return result::Error(hilti::util::fmt("value for attribute '%s' must be a string", _tag));
         }
 
+        if constexpr ( std::is_same<T, int64_t>::value ) {
```
Comment (bbannier) [inline hilti/toolchain/include/ast/attribute.h:95]: We could use `std::is_integral_v<T>` here.

Originally I was about to suggest this so this function also works with `integer::safe<T>`, but it looks like it currently doesn't implement that trait. Might be worth being more general here nevertheless.

**Review by bbannier — CHANGES_REQUESTED** [review]

File: hilti/toolchain/src/compiler/codegen/codegen.cc [hunk 9]
```diff
@@ -244,6 +244,17 @@ struct Visitor : hilti::visitor::PreOrder<void, Visitor> {
             return;
         }
 
+        int64_t priority = 0;
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply hilti/toolchain/src/compiler/codegen/codegen.cc:247]: 0 is a deliberate choice. We're mimicking semantics of Zeek's event priorities here: zero is default, and users can chose to lower or higher.

File: tests/hilti/types/function/hook-priorities.hlt (hunk 2 above)
Comment (rsmmr) [reply tests/hilti/types/function/hook-priorities.hlt:12]: See above.

File: doc/programming/parsing.rst (hunk 4 above)
Comment (rsmmr) [reply doc/programming/parsing.rst:218]: Oops, some accidental doc changes slipped in, will revert.

File: hilti/toolchain/include/ast/attribute.h (hunk 8 above)
Comment (rsmmr) [reply hilti/toolchain/include/ast/attribute.h:95]: Not sure, it's meant to get an `int64_t` ... :)

**Comment by rsmmr** [comment]
Turns out I had the sorting wrong: Mimicking Zeek's event priorities means higher priority value executes first. Fixed.

**Review by bbannier — DISMISSED** [review]

**Review by bbannier — APPROVED** [review]
I suspect that hooks with same (e.g., default) priority are executed in undetermined order will bite us in the future when writing unit tests when we forget to normalize the output with `sort`.

Do you think it would make sense to use `std::stable_sort` instead of `std::sort` to order hooks? Not sure that would promise users more than we want, but at least it should make these sorting changes unnecessary and introduce some determinism.

**Comment by rsmmr** [comment]
> I suspect that hooks with same (e.g., default) priority are executed in undetermined order will bite us in the future when writing unit tests when we forget to normalize the output with `sort`.

Yeah, unfortunately. 

> Do you think it would make sense to use `std::stable_sort` instead of `std::sort` to order hooks?

Won't help unfortunately. I don't think we can (or at least should not) rely on the internal order that the implementations are going to come into the linker.

**Review by bbannier — APPROVED** [review]
LGTM but let's wait for CI.
